libraries/chain/controller.cpp

```diff
@@ -2538,21 +2466,20 @@ struct controller_impl {
          pending.reset();
       });
 
-      auto update_pending = [&]<class fork_db_t, class bsp>(fork_db_t& fork_db, bsp& head) {
-         EOS_ASSERT( self.skip_db_sessions(s) ||
-                     db.revision() == head_block_num(), database_exception, "db revision is not on par with head block",
-                        ("db.revision()", db.revision())("controller_head_block", head_block_num())("fork_db_head_block", fork_db_head_block_num()) );
-         maybe_session session = self.skip_db_sessions(s) ? maybe_session() : maybe_session(db);
-         if constexpr (std::is_same_v<bsp, block_state_legacy_ptr>)
-            pending.emplace(std::move(session), *head, when, confirm_block_count, new_protocol_feature_activations);
-         else {
-            building_block_input bbi{ head->id(), when, head->get_scheduled_producer(when).producer_name,
-                                      new_protocol_feature_activations };
-            pending.emplace(std::move(session), *head, bbi);
-         }
-      };
+      EOS_ASSERT( self.skip_db_sessions(s) || db.revision() == head_block_num(), database_exception,
+                  "db revision is not on par with head block",
+                  ("db.revision()", db.revision())("controller_head_block", head_block_num())("fork_db_head_block", fork_db_head_block_num()) );
 
-      block_data.apply<void>(update_pending);
+      fork_db.apply_dpos<void>([&](auto& forkdb) {
+         maybe_session session = self.skip_db_sessions(s) ? maybe_session() : maybe_session(db);
+         pending.emplace(std::move(session), *forkdb.chain_head, when, confirm_block_count, new_protocol_feature_activations);
+      });
+      fork_db.apply_if<void>([&](auto& forkdb) {
+         maybe_session session = self.skip_db_sessions(s) ? maybe_session() : maybe_session(db);
+         building_block_input bbi{ forkdb.chain_head->id(), when, forkdb.chain_head->get_scheduled_producer(when).producer_name,
+                                   new_protocol_feature_activations };
+         pending.emplace(std::move(session), *forkdb.chain_head, bbi);
+      });
```

Comment:
I wonder if we should have an `apply` which takes two lambdas, one for the `dpos` code, and one for the `if` code.

Reply:
I removed `apply_if` and `apply_dpos` and started to use just the one that took two lambdas. But that makes it not at all obvious what is happening when looking at the calls in controller. I think I will add an apply that takes two but also keep the `apply_if` and `apply_legacy` (renmaing `apply_dpos` to `apply_legacy`)